### ISSUE TITLE
ACS Commons as modular library

### ISSUE THREAD

**Issue by JelleBouwmans** [Adobe-Consulting-Services/acs-aem-commons#2665]
Recently we have been getting new security vulnerabilities in ACS Commons from OWASP. Luckily, most of these were false positives or in functionalities we don’t use.

This made us think about the big number of functionalities of ACS Commons we don’t use. These functionalities do increase the attack surface with those vulnerabilities. In an ideal scenario, the ACS Commons library only includes the functionalities we use.

Reasons for us to have a library with only the functionalities we would use:
- It decreases the attack surface.
- Smart authors will not find any functionalities they should not be using.

So as feature request: The possibility for a (modular) ACS Commons library which only has the functionalities we need.

**Comment by kwin** [comment]
All features should already be opt-in features, i.e. not executable until explicitly enabled. Compare with https://github.com/Adobe-Consulting-Services/acs-aem-commons/blob/master/CONTRIBUTING.md#before-contributing-code. @JelleBouwmans Have you found any feature which were exploitable without being activated (through an explicit OSGi configuration or via an explicit reference in some client library?

**Comment by henrykuijpers** [comment]
If I look at for example the TouchUI settings, the "ACS AEM Commons"-tab, there is a lot of functionality that is active by default.

Also, there are of course a lot of services, such as servlets (Java-code) that are active by default: For example
https://github.com/Adobe-Consulting-Services/acs-aem-commons/blob/ff750e963a7ca3a1df49f4a9afa8aca4be4c63ef/bundle/src/main/java/com/adobe/acs/commons/indesign/dynamicdeckdynamo/servlets/impl/TriggerDeckDynamoServlet.java#L50
Even if Dynamic Deck Dynamo is not turned on (I don't know if there's a toggle?), this servlet is registered and active. In case Dispatcher is not configured specifically enough (or if a user accesses the author, which usually has a way more relaxed Dispatcher), this servlet could be reachable.

**Comment by kwin** [comment]
@henrykuijpers Patches to make those services only active if an OSGi configuration is provided are greatly appreciated.

**Comment by henrykuijpers** [comment]
Would it be an idea to start using the "Feature"-function of Sling for this @kwin ? https://github.com/apache/sling-org-apache-sling-feature

That way, we could define some features (such as "Dynamic Deck Dynamo"), which would make dependent services only activate when that feature is available?

**Comment by henrykuijpers** [comment]
(It still doesn't solve the OWASP findings issue btw, the dependencies are still there.)

**Comment by kwin** [comment]
Sling Features would only be the 2nd best alternative as those require an explicit guard. For OSGi components requiring a configuration this is already built in (i.e. requires less code). If multiple components depend one one configuration just chain them via explicit references: component c -> component b -> component a. Component a would require an OSGi configuration and all the other ones would also only become active after component a is active.

**Comment by davidjgonzalez** [comment]
@henrykuijpers i believe all (or at least most) of those features require direct human invocation for them to be executed.

This is in contrast to like if we turned  "Ensure Oak Indexes" was on by default (it's not) - which is a "side effect" of a non-ACS Commons specific operation.

We can discuss the risks/merits of course. 

EOD, as a community project no ones directly paid to work on, the reality is there is not enough time to manage everything as a discrete artifact.

**Comment by joerghoh** [comment]
Actually I like the idea of a more modular library, especially as we already have a set of explicitly disabled features for AEM CS. Moving them into a dedicated "AEM6" module would allow us to handle them differently.

**Comment by kwin** [comment]
I agree with @davidjgonzalez that the time spend on the effort to split up ACS Commons should rather be spent on bug fixes and features but if anyone is willing to do the modularisation I am not objecting.
